### PR TITLE
Improve IndexOfAnyAsciiSearcher ARM throughput

### LINKED ISSUE #78718: FileNameStar_NeedsEncoding_EncodedAndDecodedCorrectly test is failing in CI
https://dev.azure.com/dnceng-public/public/_build/results?buildId=91029&view=ms.vss-test-web.build-test-results-tab&runId=1918218&resultId=189964&paneView=debug

Configuration: `net7.0-Linux-Release-arm64-CoreCLR_checked-(Alpine.314.Arm64.Open)Ubuntu.1804.ArmArch.Open@mcr.microsoft.com/dotnet-buildtools/prereqs:alpine-3.14-helix-arm64v8` & `net7.0-Linux-Debug-arm64-CoreCLR_checked-(Ubuntu.1804.Arm64.Open)Ubuntu.1804.Armarch.Open@mcr.microsoft.com/dotnet-buildtools/prereqs:ubuntu-18.04-helix-arm64v8`

```
Assert.Equal() Failure
↓ (pos 4)
Expected: FileÃName.bat
Actual:   File?�Name.bat
↑ (pos 4)
   at System.Net.Http.Tests.ContentDispositionHeaderValueTest.FileNameStar_NeedsEncoding_EncodedAndDecodedCorrectly() in /_/src/libraries/System.Net.Http/tests/UnitTests/Headers/ContentDispositionHeaderValueTest.cs:line 211
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Void** arguments, Signature sig, Boolean isConstructor)
   at System.Reflection.MethodInvoker.Invoke(Object obj, IntPtr* args, BindingFlags invokeAttr) in /_/src/libraries/System.Private.CoreLib/src/System/Reflection/MethodInvoker.cs:line 59
```

```json
{
  "ErrorMessage": "Expected: FileÃName.bat",
  "BuildRetry": false
}
```

cc @dotnet/ncl 

### Report

|Build|Definition|Test|Pull Request|
|---|---|---|---|
|[91636](https://dev.azure.com/dnceng-public/public/_build/results?buildId=91636)|dotnet/runtime|[System.Net.Http.Tests.ContentDispositionHeaderValueTest.FileNameStar_NeedsEncoding_EncodedAndDecodedCorrectly](https://dev.azure.com/dnceng-public/public/_build/results?buildId=91636&view=ms.vss-test-web.build-test-results-tab&runId=1930274&resultId=189242)|dotnet/runtime#78736|
|[91625](https://dev.azure.com/dnceng-public/public/_build/results?buildId=91625)|dotnet/runtime|[System.Net.Http.Tests.ContentDispositionHeaderValueTest.FileNameStar_NeedsEncoding_EncodedAndDecodedCorrectly](https://dev.azure.com/dnceng-public/public/_build/results?buildId=91625&view=ms.vss-test-web.build-test-results-tab&runId=1929578&resultId=189489)|dotnet/runtime#78741|
|[91517](https://dev.azure.com/dnceng-public/public/_build/results?buildId=91517)|dotnet/runtime|[System.Net.Http.Tests.ContentDispositionHeaderValueTest.FileNameStar_NeedsEncoding_EncodedAndDecodedCorrectly](https://dev.azure.com/dnceng-public/public/_build/results?buildId=91517&view=ms.vss-test-web.build-test-results-tab&runId=1927280&resultId=189944)|dotnet/runtime#78738|
|[91507](https://dev.azure.com/dnceng-public/public/_build/results?buildId=91507)|dotnet/runtime|[System.Net.Http.Tests.ContentDispositionHeaderValueTest.FileNameStar_NeedsEncoding_EncodedAndDecodedCorrectly](https://dev.azure.com/dnceng-public/public/_build/results?buildId=91507&view=ms.vss-test-web.build-test-results-tab&runId=1927162&resultId=194779)|dotnet/runtime#78736|
|[91483](https://dev.azure.com/dnceng-public/public/_build/results?buildId=91483)|dotnet/runtime|[System.Net.Http.Tests.ContentDispositionHeaderValueTest.FileNameStar_NeedsEncoding_EncodedAndDecodedCorrectly](https://dev.azure.com/dnceng-public/public/_build/results?buildId=91483&view=ms.vss-test-web.build-test-results-tab&runId=1926680&resultId=189864)|dotnet/runtime#78580|
|[91457](https://dev.azure.com/dnceng-public/public/_build/results?buildId=91457)|dotnet/runtime|[System.Net.Http.Tests.ContentDispositionHeaderValueTest.FileNameStar_NeedsEncoding_EncodedAndDecodedCorrectly](https://dev.azure.com/dnceng-public/public/_build/results?buildId=91457&view=ms.vss-test-web.build-test-results-tab&runId=1926420&resultId=189419)|dotnet/runtime#78733|
|[91365](https://dev.azure.com/dnceng-public/public/_build/results?buildId=91365)|dotnet/runtime|[System.Net.Http.Tests.ContentDispositionHeaderValueTest.FileNameStar_NeedsEncoding_EncodedAndDecodedCorrectly](https://dev.azure.com/dnceng-public/public/_build/results?buildId=91365&view=ms.vss-test-web.build-test-results-tab&runId=1924916&resultId=182432)||
|[91336](https://dev.azure.com/dnceng-public/public/_build/results?buildId=91336)|dotnet/runtime|[System.Net.Http.Tests.ContentDispositionHeaderValueTest.FileNameStar_NeedsEncoding_EncodedAndDecodedCorrectly](https://dev.azure.com/dnceng-public/public/_build/results?buildId=91336&view=ms.vss-test-web.build-test-results-tab&runId=1924068&resultId=189263)|dotnet/runtime#78728|
|[91307](https://dev.azure.com/dnceng-public/public/_build/results?buildId=91307)|dotnet/runtime|[System.Net.Http.Tests.ContentDispositionHeaderValueTest.FileNameStar_NeedsEncoding_EncodedAndDecodedCorrectly](https://dev.azure.com/dnceng-public/public/_build/results?buildId=91307&view=ms.vss-test-web.build-test-results-tab&runId=1923730&resultId=189199)||
|[91215](https://dev.azure.com/dnceng-public/public/_build/results?buildId=91215)|dotnet/runtime|[System.Net.Http.Tests.ContentDispositionHeaderValueTest.FileNameStar_NeedsEncoding_EncodedAndDecodedCorrectly](https://dev.azure.com/dnceng-public/public/_build/results?buildId=91215&view=ms.vss-test-web.build-test-results-tab&runId=1922022&resultId=194374)|dotnet/runtime#78572|
|[91201](https://dev.azure.com/dnceng-public/public/_build/results?buildId=91201)|dotnet/runtime|[System.Net.Http.Tests.ContentDispositionHeaderValueTest.FileNameStar_NeedsEncoding_EncodedAndDecodedCorrectly](https://dev.azure.com/dnceng-public/public/_build/results?buildId=91201&view=ms.vss-test-web.build-test-results-tab&runId=1921642&resultId=189370)|dotnet/runtime#78709|
|[91155](https://dev.azure.com/dnceng-public/public/_build/results?buildId=91155)|dotnet/runtime|[System.Net.Http.Tests.ContentDispositionHeaderValueTest.FileNameStar_NeedsEncoding_EncodedAndDecodedCorrectly](https://dev.azure.com/dnceng-public/public/_build/results?buildId=91155&view=ms.vss-test-web.build-test-results-tab&runId=1921038&resultId=189444)|dotnet/runtime#78720|
|[90632](https://dev.azure.com/dnceng-public/public/_build/results?buildId=90632)|dotnet/runtime|[System.Net.Http.Tests.ContentDispositionHeaderValueTest.FileNameStar_NeedsEncoding_EncodedAndDecodedCorrectly](https://dev.azure.com/dnceng-public/public/_build/results?buildId=90632&view=ms.vss-test-web.build-test-results-tab&runId=1920776&resultId=189856)|dotnet/runtime#78685|
|[91127](https://dev.azure.com/dnceng-public/public/_build/results?buildId=91127)|dotnet/runtime|[System.Net.Http.Tests.ContentDispositionHeaderValueTest.FileNameStar_NeedsEncoding_EncodedAndDecodedCorrectly](https://dev.azure.com/dnceng-public/public/_build/results?buildId=91127&view=ms.vss-test-web.build-test-results-tab&runId=1919904&resultId=189277)|dotnet/runtime#78714|
|[91029](https://dev.azure.com/dnceng-public/public/_build/results?buildId=91029)|dotnet/runtime|[System.Net.Http.Tests.ContentDispositionHeaderValueTest.FileNameStar_NeedsEncoding_EncodedAndDecodedCorrectly](https://dev.azure.com/dnceng-public/public/_build/results?buildId=91029&view=ms.vss-test-web.build-test-results-tab&runId=1918218&resultId=189964)|dotnet/runtime#78703|
|[90647](https://dev.azure.com/dnceng-public/public/_build/results?buildId=90647)|dotnet/runtime|[System.Net.Http.Tests.ContentDispositionHeaderValueTest.FileNameStar_NeedsEncoding_EncodedAndDecodedCorrectly](https://dev.azure.com/dnceng-public/public/_build/results?buildId=90647&view=ms.vss-test-web.build-test-results-tab&runId=1918158&resultId=189218)|dotnet/runtime#77996|
|[91022](https://dev.azure.com/dnceng-public/public/_build/results?buildId=91022)|dotnet/runtime|[System.Net.Http.Tests.ContentDispositionHeaderValueTest.FileNameStar_NeedsEncoding_EncodedAndDecodedCorrectly](https://dev.azure.com/dnceng-public/public/_build/results?buildId=91022&view=ms.vss-test-web.build-test-results-tab&runId=1918042&resultId=189954)|dotnet/runtime#78580|
|[91010](https://dev.azure.com/dnceng-public/public/_build/results?buildId=91010)|dotnet/runtime|[System.Net.Http.Tests.ContentDispositionHeaderValueTest.FileNameStar_NeedsEncoding_EncodedAndDecodedCorrectly](https://dev.azure.com/dnceng-public/public/_build/results?buildId=91010&view=ms.vss-test-web.build-test-results-tab&runId=1917990&resultId=189261)|dotnet/runtime#78700|
|[90970](https://dev.azure.com/dnceng-public/public/_build/results?buildId=90970)|dotnet/runtime|[System.Net.Http.Tests.ContentDispositionHeaderValueTest.FileNameStar_NeedsEncoding_EncodedAndDecodedCorrectly](https://dev.azure.com/dnceng-public/public/_build/results?buildId=90970&view=ms.vss-test-web.build-test-results-tab&runId=1916944&resultId=189841)|dotnet/runtime#78702|
|[90927](https://dev.azure.com/dnceng-public/public/_build/results?buildId=90927)|dotnet/runtime|[System.Net.Http.Tests.ContentDispositionHeaderValueTest.FileNameStar_NeedsEncoding_EncodedAndDecodedCorrectly](https://dev.azure.com/dnceng-public/public/_build/results?buildId=90927&view=ms.vss-test-web.build-test-results-tab&runId=1915938&resultId=195069)|dotnet/runtime#78698|
|[90857](https://dev.azure.com/dnceng-public/public/_build/results?buildId=90857)|dotnet/runtime|[System.Net.Http.Tests.ContentDispositionHeaderValueTest.FileNameStar_NeedsEncoding_EncodedAndDecodedCorrectly](https://dev.azure.com/dnceng-public/public/_build/results?buildId=90857&view=ms.vss-test-web.build-test-results-tab&runId=1915298&resultId=189582)|dotnet/runtime#78557|
|[90830](https://dev.azure.com/dnceng-public/public/_build/results?buildId=90830)|dotnet/runtime|[System.Net.Http.Tests.ContentDispositionHeaderValueTest.FileNameStar_NeedsEncoding_EncodedAndDecodedCorrectly](https://dev.azure.com/dnceng-public/public/_build/results?buildId=90830&view=ms.vss-test-web.build-test-results-tab&runId=1914264&resultId=189408)|dotnet/runtime#78020|
|[90801](https://dev.azure.com/dnceng-public/public/_build/results?buildId=90801)|dotnet/runtime|[System.Net.Http.Tests.ContentDispositionHeaderValueTest.FileNameStar_NeedsEncoding_EncodedAndDecodedCorrectly](https://dev.azure.com/dnceng-public/public/_build/results?buildId=90801&view=ms.vss-test-web.build-test-results-tab&runId=1913512&resultId=189701)|dotnet/runtime#78678|
|[90662](https://dev.azure.com/dnceng-public/public/_build/results?buildId=90662)|dotnet/runtime|[System.Net.Http.Tests.ContentDispositionHeaderValueTest.FileNameStar_NeedsEncoding_EncodedAndDecodedCorrectly](https://dev.azure.com/dnceng-public/public/_build/results?buildId=90662&view=ms.vss-test-web.build-test-results-tab&runId=1909942&resultId=188742)||
|[90652](https://dev.azure.com/dnceng-public/public/_build/results?buildId=90652)|dotnet/runtime|[System.Net.Http.Tests.ContentDispositionHeaderValueTest.FileNameStar_NeedsEncoding_EncodedAndDecodedCorrectly](https://dev.azure.com/dnceng-public/public/_build/results?buildId=90652&view=ms.vss-test-web.build-test-results-tab&runId=1909438&resultId=195056)|dotnet/runtime#78688|
|[90641](https://dev.azure.com/dnceng-public/public/_build/results?buildId=90641)|dotnet/runtime|[System.Net.Http.Tests.ContentDispositionHeaderValueTest.FileNameStar_NeedsEncoding_EncodedAndDecodedCorrectly](https://dev.azure.com/dnceng-public/public/_build/results?buildId=90641&view=ms.vss-test-web.build-test-results-tab&runId=1909182&resultId=189512)||
|[90620](https://dev.azure.com/dnceng-public/public/_build/results?buildId=90620)|dotnet/runtime|[System.Net.Http.Tests.ContentDispositionHeaderValueTest.FileNameStar_NeedsEncoding_EncodedAndDecodedCorrectly](https://dev.azure.com/dnceng-public/public/_build/results?buildId=90620&view=ms.vss-test-web.build-test-results-tab&runId=1908802&resultId=189979)||
|[90558](https://dev.azure.com/dnceng-public/public/_build/results?buildId=90558)|dotnet/runtime|[System.Net.Http.Tests.ContentDispositionHeaderValueTest.FileNameStar_NeedsEncoding_EncodedAndDecodedCorrectly](https://dev.azure.com/dnceng-public/public/_build/results?buildId=90558&view=ms.vss-test-web.build-test-results-tab&runId=1908022&resultId=189625)|dotnet/runtime#78680|
#### Summary
|24-Hour Hit Count|7-Day Hit Count|1-Month Count|
|---|---|---|
|28|28|28|

### PR DESCRIPTION
**Pull request by MihaZupan**
Also solves an edge-case bug for byte overloads on ARM64 where we'd get false positives (negatives for Except) when haystack values were exactly 128 above a value in the needle.
Closes #78718

Before
|             Method | Length |       Needle |      Mean |     Error |
|------------------- |------- |------------- |----------:|----------:|
| IndexOfAny128_Char | 100000 | AlphaNumeric |  18.27 us |  0.035 us |
| IndexOfAny128_Byte | 100000 | AlphaNumeric |  11.37 us |  0.032 us |

After
|             Method | Length |       Needle |      Mean |     Error |
|------------------- |------- |------------- |----------:|----------:|
| IndexOfAny128_Char | 100000 | AlphaNumeric | 16.363 us | 0.3711 us |
| IndexOfAny128_Byte | 100000 | AlphaNumeric |  9.621 us | 0.0884 us |

### CODE REVIEW

File: src/libraries/System.Memory/tests/Span/IndexOfAny.byte.cs [hunk 1]
```diff
@@ -553,6 +553,51 @@ static int IndexOfAnyReferenceImpl(ReadOnlySpan<byte> searchSpace, ReadOnlySpan<
...
+            Assert.Equal(needleContainsZero ? 8 : 6, haystackWithZeroes.LastIndexOfAny(needle));
+            Assert.Equal(needleContainsZero ? 7 : 8, haystackWithZeroes.LastIndexOfAnyExcept(needle));
+
+            Span<byte> haystackWithOffsetNeedle = new byte[100];
+            for (int i = 0; i < haystackWithOffsetNeedle.Length; i++)
+            {
+                haystackWithOffsetNeedle[i] = (byte)(128 + needleValues[i % needleValues.Length]);
+            }
+
+            Assert.Equal(-1, haystackWithOffsetNeedle.IndexOfAny(needle));
```
Comment (MihaZupan) [inline src/libraries/System.Memory/tests/Span/IndexOfAny.byte.cs:584]: The test would blow up here on ARM before this change

File: src/libraries/System.Private.CoreLib/src/System/IndexOfAnyValues/IndexOfAnyAsciiSearcher.cs [hunk 2]
```diff
@@ -490,26 +490,23 @@ internal static int LastIndexOfAnyVectorized<TNegator>(ref byte searchSpace, int
...
-            if (AdvSimd.Arm64.IsSupported || TOptimizations.NeedleContainsZero)
+            if (TOptimizations.NeedleContainsZero)
             {
+                Debug.Assert(Sse2.IsSupported);
                 Vector128<short> ascii0 = Vector128.LessThan(source0.AsUInt16(), Vector128.Create((ushort)128)).AsInt16();
                 Vector128<short> ascii1 = Vector128.LessThan(source1.AsUInt16(), Vector128.Create((ushort)128)).AsInt16();
-                Vector128<byte> ascii = Sse2.IsSupported
-                    ? Sse2.PackSignedSaturate(ascii0, ascii1).AsByte()
-                    : AdvSimd.Arm64.UnzipEven(ascii0.AsByte(), ascii1.AsByte());
+                Vector128<byte> ascii = Sse2.PackSignedSaturate(ascii0, ascii1).AsByte();
```
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/IndexOfAnyValues/IndexOfAnyAsciiSearcher.cs:509]: Nit: does the ascii local temporary help with anything?

**Review by stephentoub — APPROVED** [review]

File: src/libraries/System.Private.CoreLib/src/System/IndexOfAnyValues/IndexOfAnyAsciiSearcher.cs (hunk 2 above)
Comment (MihaZupan) [reply src/libraries/System.Private.CoreLib/src/System/IndexOfAnyValues/IndexOfAnyAsciiSearcher.cs:509]: I don't think so, I just found it a bit more readable when initially writing this logic

**Review by EgorBo — APPROVED** [review]

**Pull request merged**